src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java

```diff
@@ -0,0 +1,215 @@
+
+        int code = proc.join();
+
+        if (code != 0) {
+            throw new MatlabExecutionException(code);
+        }
+    }
+
+    /**
+     * Redirects stdout.
+     *
+     * @param out the OutputStream to write to
+     */
+    public void redirectStdOut(OutputStream out) {
+        this.stdOut = out;
+    }
+
+    /**
+     * Adds an environment variable.
+     *
+     * @param key the environment variable name
+     * @param value the environment variable value
+     */
+    public void addEnvironmentVariable(String key, String value) {
+        additionalEnvVars.put(key, value);
+    }
+
+    public EnvVars getEnvVars() {
+        EnvVars env = new EnvVars(this.params.getEnvVars());
+        env.putAll(additionalEnvVars); 
+        return env;
+    }
+
+    /**
+     * Copies a resource into the temporary folder.
+     *
+     * @param sourceFile the name of a resource the class loader can find.
+     * @param targetFile the name of the file to create in the temp folder.
+     * @return the FilePath to the new location in the temp folder.
+     */
+    public FilePath copyFileToTempFolder(String sourceFile, String targetFile) throws IOException, InterruptedException {
+        final ClassLoader classLoader = getClass().getClassLoader();
+        FilePath targetFilePath = new FilePath(this.tempFolder, targetFile);
+        InputStream in = classLoader.getResourceAsStream(sourceFile);
+        targetFilePath.copyFrom(in);
+        targetFilePath.chmod(0755);
+
+        return targetFilePath;
+    }
+
+    public FilePath getTempFolder() {
+        return tempFolder;
+    }
+
+    /**
+     * Creates a file with the specified content in the temporary folder.
+     *
+     * Additionally, the file content will be prefixed with a statement returning to the MATLAB starting folder.
+     *
+     * @param content string that represents the content of the file.
+     * @return the FilePath to the script file that is created.
+     */
+    protected FilePath createFileWithContent(String content) throws IOException, InterruptedException {
+        String fileName = "script_" + RandomStringUtils.randomAlphanumeric(8) + ".m";
+        FilePath scriptFile = new FilePath(this.tempFolder, fileName);
+
+        String expandedContent = getEnvVars().expand(content);
+        String finalContent = "cd(getenv('MW_ORIG_WORKING_FOLDER'));\n"
+            + expandedContent;
+
+        this.params.getTaskListener().getLogger()
+            .println("Generating MATLAB script with content:\n" + expandedContent + "\n\n");
+
+        scriptFile.write(finalContent, "UTF-8");
+
+        return scriptFile;
+    }
+
+    /**
+     * Copies platform specific runner file into the temporary folder.
+     *
+     * @return the FilePath to the runner executable
+     */
+    protected FilePath prepareRunnerExecutable() throws IOException, InterruptedException {
+        Launcher launcher = this.params.getLauncher();
+        if (launcher.isUnix()) {
```

Comment:
I believe that this is necessary to get the correct executable for the platform the plugin is running on. We'll need to check the operating system and differentiate between Mac and Linux at some point, and this seemed like the natural place to do so.

Getting the destination folder name is far less important than getting the correct source folder location. In fact I believe that the destination location is the same for both branches, we just copy a different executable to that destination.

This is also identical to how we're differentiating between Mac and Linux in the current iteration of the plugin. If you have another suggestion that you believe would be an improvement definitely let me know, because ideally we wouldn't need to spawn an extra process.

Reply:
is `run-matlab-command` different for MAc & linux ?

Reply:
Yep!

![image](https://github.com/mathworks/jenkins-matlab-plugin/assets/122639681/d0ad5ad1-47e9-42a0-8b27-f374cea6bb3f)

Reply:
Ahh, makes sense in that case.